reco/HoughTransformation/ATHoughSpaceCircle.cc

```diff
@@ -90,338 +90,352 @@ TH2F* ATHoughSpaceCircle::GetHoughSpace(TString ProjPlane)
+
+        TVector3 const position = hit.GetPosition();
+        TVector3 const position_forw = hit_forw.GetPosition();
+        TVector3 const position_next = hit_next.GetPosition();
+
+        TVector3 const deltaPosNext = position - position_next;
+        Double_t const hitdist = TMath::Sqrt(deltaPosNext * deltaPosNext);
+
+        for (Int_t itheta = 0; itheta < 1023; ++itheta)
+        {
+            TVector3 const deltaPosForw = position - position_forw;
+            Float_t const angle = TMath::Pi() * (static_cast<Float_t>(itheta) / 1023.0f);
+            Float_t const d0_XY_inv = 2.0f * (deltaPosForw.Y() * TMath::Sin(angle) + deltaPosForw.X() * TMath::Cos(angle))
+                / ((TMath::Power(position.Y(), 2) - TMath::Power(position_forw.Y(), 2))
+                 + (TMath::Power(position.X(), 2) - TMath::Power(position_forw.X(), 2)));
+
+            HistHoughXY->Fill(angle, 1.0f / d0_XY_inv); // TODO: flip calculation so we do not need to do the inversion here
+        }
+    }
+}
 
-                     if(fClusteredTBHits_forw.size()==0){// If forward hit is empty scan until the end of the vector or until the next non-zero solution
+void ATHoughSpaceCircle::Clustering(ATEvent *event)
+{
+    Int_t const nHits = event->GetNumHits();
 
+    // Double_t IniPosZ = 0.0;
+    Int_t IniTSBuff = 0;
+    Double_t IniRadiusBuff = 0.0;
+    Double_t IniPhiBuff = 0.0;
 
-                          while(fClusteredTBHits_forw.size()==0  &&  fIniTS-iHitTB-1-TB_Scan>0){ //TODO: Use a lambda function for this mess
-                              TB_Scan++;
-                              fClusteredTBHits_forw=min->GetTBHitArray(fIniTS-iHitTB-1-TB_Scan,fClusteredHits);
-                              //std::cout<<TB_Scan<<std::endl;
-                              //std::cout<<fClusteredTBHits_forw.size()<<std::endl;
-                          }
+    for (Int_t iHit = 0; iHit < (nHits - 1); ++iHit)
+    {
+        // TODO: better leave event const
+        ATHit &hit = event->GetHitArray()->at(iHit);
+        TVector3 const position = hit.GetPosition();
+        Double_t const geo_dist = TMath::Abs(
+            TMath::Cos(fHoughLinePar.first) * fPhi->at(iHit) * fRadius->at(iHit)
+            + TMath::Sin(fHoughLinePar.first) * hit.GetTimeStamp()
+            - fHoughLinePar.second
+        );
+        Double_t const hit_dist = TMath::Sqrt(
+            TMath::Power((fPhi->at(iHit) * fRadius->at(iHit) - IniPhiBuff * IniRadiusBuff), 2)
+            + TMath::Power((hit.GetTimeStamp() - IniTSBuff), 2)
+        );
+        // if (hit_dist < 50) std::cout << " Hist Dist pre:" << hit_dist << " Geo Dist: " << geo_dist << std::endl;
+
+        if (geo_dist < 10.0)
+        {
+            IniTSBuff = hit.GetTimeStamp();
+            IniRadiusBuff = fRadius->at(iHit);
+            IniPhiBuff = fPhi->at(iHit);
+
+            if (hit_dist < 10.0)
+            {
+                fClusteredHits->push_back(&hit); // TODO: why pointer? pointer to stack? why not copy?
```

Comment:
Regarding pointers, we should stop using new and delete entirely.
They only belong in data-structure-classes (like vector, list, etc.)

And you do not need to pass a pointer every time a big object is returned. I'm a huge fan of `const &` if you didn't see already. And since move-constructors are a thing, a normal "copy" of a vector, for example, isn't a problem either.
And then, what you are left with, are most of the time simple pointers which just help you find that element in a array of your own object. And that kind of pointer doesn't need to be managed.